guest-tools/tdvirsh

```diff
@@ -0,0 +1,195 @@
+  print;
+        }
+  ' ${xml_template_path} > ${WORKDIR_PATH}/${DOMAIN_PREFIX}.xml
+}
+
+boot_vm() {
+    virsh define ${WORKDIR_PATH}/${DOMAIN_PREFIX}.xml >/dev/null
+    created_domain=${DOMAIN_PREFIX}-$(virsh domuuid ${DOMAIN_PREFIX})
+    mv ${WORKDIR_PATH}/{${DOMAIN_PREFIX}.xml,${created_domain}.xml} &>/dev/null || true
+    virsh domrename ${DOMAIN_PREFIX} ${created_domain} >/dev/null
+    virsh start ${created_domain} >/dev/null
+}
+
+destroy() {
+    local domain_to_destroy="${1}"
+    local qcow2_overlay_path
+    qcow2_overlay_path=$(virsh dumpxml ${domain_to_destroy} |
+        grep -o "${WORKDIR_PATH}\/overlay\.[A-Za-z0-9]*\.qcow2")
+
+    echo "Destroying domain ${domain_to_destroy}"
+
+    virsh shutdown ${domain_to_destroy} &>/dev/null
+    virsh shutdown --domain ${domain_to_destroy} &>/dev/null
+
+    echo "Waiting for VM to shutdown ..."
+    sleep 5
+
+    virsh destroy ${domain_to_destroy} &>/dev/null
+    virsh destroy --domain ${domain_to_destroy} &>/dev/null
+    virsh undefine ${domain_to_destroy} &>/dev/null
+
+    rm -f ${qcow2_overlay_path} ${WORKDIR_PATH}/${domain_to_destroy}.xml
+}
+
+clean_all() {
+    for domain_to_clean in $(virsh list --all --name | grep ${DOMAIN_PREFIX}); do
+        destroy ${domain_to_clean}
+    done
+    rm -rf ${WORKDIR_PATH}
+}
+
+print_all() {
+    while read -r line
+    do
+        local extra_info=""
+        local td_domain=$(echo $line | grep -oP "${DOMAIN_PREFIX}-[^ ]+")
+        if [ ! -z "$td_domain" ]; then
+            host_port=$(
+                virsh \
+                    qemu-monitor-command ${td_domain} \
+                    --hmp info usernet 2>&1 |
+                    awk '/HOST_FORWARD/ {print $4}'
+                     )
+            guest_cid=$(
+                virsh \
+                    qemu-monitor-command ${td_domain} \
+                    --hmp info qtree 2>&1 |
+                    awk '/guest-cid/ {print $3}'
+                     )
+            extra_info="(ssh:$host_port, cid:${guest_cid})"
+        fi
+        echo "$line $extra_info"
+    done < <(virsh "$@")
+}
+
+run_td() {
+    echo "Create and run new TD ..."
+    echo "---"
+    create_overlay_image
+    create_domain_xml
+    boot_vm
+    print_all dominfo "${created_domain}"
+}
+
+parse_params() {
+    while :; do
+        case "${1-}" in
+        "" | -h | --help)
+            usage
+            exit 0
+            ;;
+        new)
+            run_td
+            exit 0
+            ;;
+        list)
+            # NB: for the sake of simplicity, just force the list command to --all
+            # do not handle other arguments allowed by virsh
+            print_all "list --all"
+            exit 0
+            ;;
+        delete)
+            domain_to_clean="${2-}"
+            if [ "${domain_to_clean}" == "all" ]; then
+                clean_all
+            else
+                if [[ ! "${domain_to_clean}" =~ ^${DOMAIN_PREFIX}-[^\ ]+$ ]]; then
```

Comment:
This fails if the UUID is off by a digit (which could happen from a bad copy-paste) and then shows some misleading output from `destroy()`:

```
$ ./tdvirsh list
Id   Name                                            State 
--------------------------------------------------------------- 
2    td_guest-6d7f26a1-168e-4813-aa26-5127eeec388d   running (ssh:33827, cid:4)
 
$ ./tdvirsh delete td_guest-6d7f26a1-168e-4813-aa26-5127eeec388
error: failed to get domain 'td_guest-6d7f26a1-168e-4813-aa26-5127eeec388'
Destroying domain td_guest-6d7f26a1-168e-4813-aa26-5127eeec388
Waiting for VM to shutdown ...
```

We could use `set -e` at the top of the `destroy()` function so it exits on the first command using an invalid domain name.

Reply:
Adding `set -e` will give us other issues like we will stop to clean up the domain and its associated data if the domain is already shutdown or destroyed ... i propose to to sanity check to make sure the domain existsn if it does, we will do all the operations no matter the outcome of each individual operation